modules/xphoto/test/test_grayworld.cpp

```diff
@@ -0,0 +1,86 @@
+#include "test_precomp.hpp"
+
+namespace cvtest {
+
+    using namespace cv;
+
+    void ref_autowbGrayworld(InputArray _src, OutputArray _dst, const float thresh)
+    {
+        Mat src = _src.getMat();
+
+        _dst.create(src.size(), src.type());
+        Mat dst = _dst.getMat();
+
+        int width  = src.cols,
+            height = src.rows,
+            N      = width*height,
+            N3     = N*3;
+
+        // Calculate sum of pixel values of each channel
+        const uchar* src_data = src.ptr<uchar>(0);
+        unsigned long sum1 = 0, sum2 = 0, sum3 = 0;
+        int i = 0;
+        double minRGB, maxRGB, satur;
```

Comment:
even in the reference implementation I would suggest to use float's instead of double's.